src/Symfony/Component/HttpKernel/Kernel.php

```diff
@@ -110,6 +110,9 @@ public function boot()
     {
         if (true === $this->booted) {
             if (!$this->requestStackSize && $this->resetServices) {
+                if (method_exists(Request::class, 'resetFormats')) {
```

Comment:
I don't like that this leaks in the Kernel.
Couldn't we move this to RequestStack and implement the ResetInterface there? (we might have discussed it in Vienna but I'm not anymore on the conclusion on this possibility)

Reply:
Indeed having everything in the same component is better.
Tools that doesn't use `RequestStack` (as we discussed) don't use HttpKernel either anyway.

Reply:
That's adding a new public method to `RequestStack`, so should that be marked as `@internal` to go in 6.4?

Reply:
I wonder if that's necessary or if we couldn't handle it inside `RequestStack::push()` when the stack is empty?

Reply:
@fabpot you know I'm not sure I love implementing `ResetInterface` on `RequestStack`. Consider it'll look like:

```php
    /**
     * Resets stateful request properties.
     * @return void
     * @internal
     */
    public function reset(): void
    {
        Request::resetFormats();
    }
```

And yet, is this what you would expect "resetting" the request stack to do? It doesn't seem like it semantically makes sense in context of where it's placed.

Likewise I wouldn't put it in `push()` with a check for an empty stack, because it seems like it's polluting the behaviour of `push` and violating single responsibility.

Up to you (obviously) but the `HttpKernel` seems like the thing which should be responsible for doing this to me. Can we not just merge the existing approach with `@internal` and get rid of the `method_exists` call? It's not as if `Kernel` doesn't already rely on certain things existing in the `Request` class; it's not a decoupled component.

Reply:
We have to keep the `method_exists` check because versions can be mixed. HttpKernel can be used with an older version of HttpFoundation.

Otherwise, both approaches work for me.

Reply:
Fair enough in respect of `method_exists` but would be good to get an answer from Fabien on the main question around this solution versus `RequestStack`. Use of the `ResetInterface` doesn't feel very intuitive to me there as per my other comment but I'm happy to go with whatever someone is willing to accept for merge.